File: android/bridge/src/main/java/ipfs/gomobile/android/InputStreamToGo.java
package ipfs.gomobile.android;

import java.io.InputStream;

final class InputStreamToGo implements core.Reader {
    private final InputStream inputStream;

    InputStreamToGo(InputStream inputStream) {
        this.inputStream = inputStream;
    }

    public long read(byte[] p) throws Exception {
        long r = inputStream.read(p);

        if (r == -1) {
            inputStream.close(); // Auto-close inputStream when EOF is reached
            throw new Exception("EOF");
        }

        return r;
    }
}

File: ios/Bridge/GomobileIPFS/Sources/InputStreamToGo.swift
//
//  InputStreamToGo.swift
//  GomobileIPFS
//
//  Created by CI Agent on 20/01/2020.
//

import Foundation
import Core

public class InputStreamToGo: NSObject, CoreReaderProtocol {
    private var inputStream: InputStream

    init(_ inputStream: InputStream) {
        self.inputStream = inputStream
        self.inputStream.open()
    }

    public func read(_ buffer: Data?, n len: UnsafeMutablePointer<Int>?) throws {
        var read: Int

        let bytes = UnsafeMutablePointer<UInt8>(OpaquePointer((buffer! as NSData).bytes))
        read = self.inputStream.read(bytes, maxLength: buffer!.count)
        len?.initialize(to: read)

        if read == 0 && self.inputStream.streamStatus == .atEnd {
            self.inputStream.close()
            throw NSError(domain: "", code: 0, userInfo: ["NSLocalizedDescription": "EOF"])
        }
    }
}
